### ISSUE TITLE
The "masknet" in DCCRN

### ISSUE THREAD

**Issue by Tigerwww** [asteroid-team/asteroid#359]
Hi, I'm training DCCRN for denoising by #278

> You can just use DNS-chanllenge's baseline recipe with several modifications.
> 
> ```
> def make_model_and_optimizer(conf):
>     """Function to define the model and optimizer for a config dictionary.
>     Args:
>         conf: Dictionary containing the output of hierachical argparse.
>     Returns:
>         model, optimizer.
>     The main goal of this function is to make reloading for resuming
>     and evaluation very simple.
>     """
>     # Make the complete model
>     model = DCCRNet(**conf["filterbank"], **conf["masknet"])
>     # Define optimizer of this model
>     optimizer = make_optimizer(model.parameters(), **conf["optim"])
>     return model, optimizer
> 
> 
> class SimpleSystem(System):
>     def common_step(self, batch, batch_nb, train):
>         mixture, speech, noise = batch
>         estimate = self(mixture.unsqueeze(1))
>         loss = self.loss_func(estimate, speech)
>         return loss
> ```

When loading the `conf["masknet"]` to DCCRN, it says "unexpected keyword '`hidden_size`'". Would you please provide the conf setting of masknet for DCCRN?

Now I just load the architecture of DCCRN by: 
`model = DCCRNet(stft_kernel_size=512, sample_rate=8000.0, stft_stride=256, architecture="DCCRN-CL")`
The val_loss just stoped around -12(should be perfect at -97), do I miss someting here? I printed DCCRN's structure and there exists the masknet strucutre. So there is no need to set config of "masknet"?

And you did some modification yesterday and I update my local Asteroid, however, when training DCCRN, the loss will be NaN after several epochs.`(loss_func = PITLossWrapper(SingleSrcNegSDR("snr"),pit_from='pw_pt'`)) Is there something wrong or because of my non-config of masknet?

**Comment by jonashaag** [comment]
Hi! Yes it’s a it confusing currently. Wait for the 0.4 release and then do a training from scratch.

Can you explain why you expect the model to get 97 dab performance? Are you training on identity?

The way you initialize the model looks good.

**Comment by Tigerwww** [comment]
Thank you very much for your reply!

I have to say that I do not major in Speech Separation or Speech Enhancement,  so I might make many mistakes. I'm very glad if you could give me precious instructions.
> Can you explain why you expect the model to get 97 dB performance? Are you training on identity?

I want to train DCCRN for denoising tasks, so I prepared my parallel dataset like`[noisy speech, clean speech]`.
If I directly use `loss_func = PITLossWrapper(SingleSrcNegSDR("snr"))` to calculate `[clean speech, clean speech]`, the result would be around -97. So that's why I said above.

**Comment by jonashaag** [comment]
-12 dB is a good result for denoising depending on your dataset :)

**Comment by Tigerwww** [comment]
> -12 dB is a good result for denoising depending on your dataset :)

But the denoising result is bad. My dataset is a single-channel English dataset mixed with background noise with SNR 5. Maybe there's something wrong with the settings and I'm still checking them.

**Comment by jonashaag** [comment]
What Git commit of Asteroid are you using to train the model? There were some mistakes in the initial implementation. Can you upload an example result?

**Comment by Tigerwww** [comment]
I use Asteroid 0.4.0rc0, just before you did modifications yesterday(Merge `asteroid.filterbanks` to `asteroid-filterbanks`).

I uploaded the samples here. It's from the training set with `train_loss = -11.9` while the `val_loss` has stopped ad -9.3 for a while.

Input.wav: from the training sample         
Denoised.wav: output by DCCRN
[samples.zip](https://github.com/mpariente/asteroid/files/5602484/samples.zip)

**Comment by jonashaag** [comment]
Funny that it learns anything at all @mpariente :-D

Yeah so please retry when 0.4 is released.

**Comment by mpariente** [comment]
0.4.0 was released yesterday, can you re-install and try it again please?

**Comment by Tigerwww** [comment]
> 0.4.0 was released yesterday, can you re-install and try it again please?

Sorry to reply this late. Very appreciate for informing. I'll try it tonight.

**Comment by Tigerwww** [comment]
@jonashaag 
@mpariente 
I re-trained DCCRN on my own dataset and it works! There still some noise left but this might be related to the small data set I used.

I uploaded some results here.
[temp_samples.zip](https://github.com/asteroid-team/asteroid/files/5643548/temp_samples.zip)

Thank you very for fixing it fast!

**Comment by Tigerwww** [comment]
> All examples have the same noise, is that the only noise you trained on?

Actually The noise I used from a dataset that was recorded at the bus station/street/subway station/restaurants.  I randomly chose and added the noise to the clean audio to build the training dataset.





> Mind sharing the pretrained model? (We can upload it for you if you don't want to bother with that, while crediting you.)

I don't mind this. But I think this pretrained model is not good, due to the limited training data. Now I'm downloading DNS challenge and LibriMix to train DCCRN.

**Comment by mpariente** [comment]
> I re-trained DCCRN on my own dataset and it works! 

Very cool news, happy that we fixed it ! :wink: 
Thanks for letting us know.

**Comment by jonashaag** [comment]
@Tigerwww did you ever reach good quality with DCCRN? :)

**Comment by Tigerwww** [comment]
> @Tigerwww did you ever reach good quality with DCCRN? :)

Very sorry that I reply such late. 

It's a little bit busy recently.  But everything is fine, after solving some problems when processing the DNS dataset and training settings. And I retrain the DCCRN this morning due to the careless mistake I made. And I checked the eval result even if it was just trained 4 epochs,  the denoising is effective(Please check the attached samples). 

The training is kind of time-consuming:  around 1 hour 10 mins for 1 epoch. And to train fast, I just used half amount of DNS dataset.  And I will upload the model and eval samples after the training.  Also, after check the final result,  I'll train DCCRN on the full DNS dataset.

BTW,  Merry Christmas and Happy Holidays！(My local time is 24th Dec. already)

PS: About the samples:
\*.wav : noisy input
de_\*.wav: denoised result
[temp_samples.zip](https://github.com/asteroid-team/asteroid/files/5738461/temp_samples.zip)

**Comment by jonashaag** [comment]
Thanks for the response, and happy holidays :)

**Comment by Tigerwww** [comment]
@jonashaag 
Happy New Year!

I have finished the training DCCRN on the full DNS dataset. 

The eval result is pretty good! But there's a small problem: The volume(Power) of the denoised audio would decrease a little compared to its original one. I think this might be related to the pre-process settings of the DNS dataset. I didn't use the official recipe to process DNS, for it keeps reporting warnings. So I wrote the recipe on my own. There's a parameter about `target level`, which the official recipe seems to cut the `target level` into different parts, while I chose to uniformly sample this value between the lower and upper of `target level` for each audio sample. Also, I re-process the dataset with no `target level` and the corresponding DCCRN is under training.

Some training settings:

Call for DCCRN: 
`model = DCCRNet(stft_kernel_size=512, sample_rate=8000.0, stft_stride=256, architecture="DCCRN-CL")
`
Sampling rate: 8000kHz
Epochs: 200
Batch_size: 128  
Half_lr: True
Early_stop: True
Optimizer: adam
lr: 0.001
Weight_decay: 0.

Please see this [link](https://drive.google.com/drive/folders/1Tf2saj74YvbRp30Rp58S-wTceGvlF-ik?usp=sharing) about the model and denoised sample.
Due to the eval dataset is small(Just 10k utterances), I attached both the best-val model and the last one(The last one stopped at the 100th epoch).

**Comment by mpariente** [comment]
Thanks for the attachment, it does sound good!

> But there's a small problem: The volume(Power) of the denoised audio would decrease a little compared to its original one. 

Which loss do you use for training? 
SI-SDR has this effect of being invariant to the scale so that the output scale doesn't match the input one.

**Comment by Tigerwww** [comment]
> Which loss do you use for training?
> SI-SDR has this effect of being invariant to the scale so that the output scale doesn't match the input one.

Yes, I used Si-SDR loss. Thank you for instructing me about this! I'll try SNR loss.

**Comment by jonashaag** [comment]
You can just fix the volume in as a post processing step.

**Comment by mpariente** [comment]
>  I'll try SNR loss.

You can also try SD-SDR (scale dependant SDR) which should have the advantage of SI-SDR (projection are orthogonal) but there is a term that is sensible to scaling. 

> You can just fix the volume in as a post processing step.

Yes, but it's not a very good method.

**Comment by mpariente** [comment]
This seems solved, closing.
